### PR TITLE
Add merge sections backend functionality, refactor handlers (#193, #198)

### LINKED ISSUE #193: refactor API `enrollUsers()` method to accept list of `sectionUsers`
Minor API design question: should `enrollUsers` accept the list of `sectionUsers`? Then it could be re-used more effectively. I guess at that point, it might be more of a `SectionHandler` object that could do other section-related operations. Maybe these are niceties that are not necessary at this point, but it could get us closer to a `cavnvasapi`-like interface.

_Originally posted by @ssciolla in https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/165#r702143192_

---

It's a better design than the one used for the creation section support code, which was the pattern for the bulk enrollment support code.  It would allow for the handler's methods to be used again within the same invocation without calling the constructor again.

### LINKED ISSUE #198: Create backend API endpoints, Canvas request processes for Merge Sections feature
This issue represents the backend work required to make the Merge Sections feature work. We will need slightly different mechanisms for instructors and admins to get sections:

Getting sections
- for instructors, simply get the sections they teach in a given term (probably `/courses/` endpoint)
- for admins, get courses in accounts they administer in a given term, with optional filtering by instructor and course name
- Course relationship should somehow be traverse-able, maybe by nesting the sections with their course

Merge/un-merge sections
- This should be more straightforward; section IDs will be needed

This work will almost certainly involve new scopes and refactoring in order  re-use existing processes.

Parent issue is #16, and sibling issue is #182.

### PR DESCRIPTION
**Pull request by ssciolla**
This PR introduces the necessary server routes and Canvas API calls to meet the requirements of the Merge Sections feature, including section fetching for instructors and admins (including filters for the latter). It also contains a significant refactor of separate API handler classes, started by @pushyamig and @lsloan, to be organized around Canvas entities and Canvas URL stems. The PR aims to resolve #193 and #198.

To Do
- [x] Add course section getting routes for instructor and admins and associated Canvas calls
- [x] Refactor `CreateSectionsApiHandler` as `CourseApiHandler`
- [x] Refactor `EnrollSectionUsersApiHandler` as `SectionApiHandler`
- [x] Add merge and un-merge routes to controller
- [x] Add service methods for above, or create SectionHandler
- [x] Re-name `privilegeLevelOneScopes` to `canvasScopes` (seems like everyone can access all routes, they just may not get data back)

Future Work?
- [ ] Consider limiting access to `/admin/sections/` by role from LTI launch (probably a separate issue)

### CODE REVIEW

File: ccm_web/server/src/api/api.controller.ts
Comment (chrisrrowland) [inline ccm_web/server/src/api/api.controller.ts:122]: Pretty sure there is no unmerge requirement for CCM, was this  just out of completeness?

Comment (ssciolla) [reply ccm_web/server/src/api/api.controller.ts:122]: It's listed in issue #16, and I see it in the admin version wireframes ("unmerge" button on success screen?).

File: ccm_web/server/src/api/api.controller.ts
Comment (chrisrrowland) [inline ccm_web/server/src/api/api.controller.ts:110]: What do you think about 
/course/:id/mergesections
with a body of section ids?

That would allow multiple sections to be merged into the target course in a single call.

Its possible I still don't have my  head wrapped around what is  actually going on though.

File: ccm_web/server/src/api/api.controller.ts
Comment (chrisrrowland) [reply ccm_web/server/src/api/api.controller.ts:122]: You're right.. I was remembering talking about no umerging in a zoom call but now I can't remember the whole context.  Carry on :)

File: ccm_web/server/src/api/api.controller.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.controller.ts:110]: Yeah, I could do this, seems like it would simplify things for you. Does it make sense for unmerge to be bulk too?

Comment (ssciolla) [reply ccm_web/server/src/api/api.controller.ts:110]: I have to think more about how unmerge bulk would look (like the route and parameters, body, etc.). You don't need the `courseId` for this.

Comment (chrisrrowland) [reply ccm_web/server/src/api/api.controller.ts:110]: > Yeah, I could do this, seems like it would simplify things for you. Does it make sense for unmerge to be bulk too?

In the UI it looks like unmerge is intended to be section by section, but would it be just the same on the backend if it accepted an array?

Comment (ssciolla) [reply ccm_web/server/src/api/api.controller.ts:110]: Yeah, I guess there is no reason we can't do this in a bulk way, but in practice, do we really need that functionality yet?

Comment (ssciolla) [reply ccm_web/server/src/api/api.controller.ts]: @chrisrrowland, okay, are you happier with these changes? https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/199/commits/0c009a3859045bd02cafd1ef114fca6263df892b and https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/199/commits/fa17d17b2013d892f7a0f29c8299719b31cfa55e

File: ccm_web/server/src/api/api.controller.ts
Comment (pushyamig) [inline ccm_web/server/src/api/api.controller.ts:89]: I don't think the UI know which terms this course belongs to. Term info mostly come into play Merge section feature. Either you have to send that info to FE through globals or store it into session or both.

Comment (ssciolla) [reply ccm_web/server/src/api/api.controller.ts:89]: I'm sending it through the course endpoint (see changes to `/course/:id` endpoint and `CanvasCourseBase`, which the frontend gets on load and can pass to the feature. It seemed better to me to get it that way, but I could do through the launch/session.

Comment (ssciolla) [reply ccm_web/server/src/api/api.controller.ts:89]: I *think* doing it this way may circumvent some of the issues related to #113, because the React application should still have in local state which course was sent in `globals` after the launch, and then be able to pull the latest term info even if session data changed.

Comment (pushyamig) [reply ccm_web/server/src/api/api.controller.ts:89]: That would be yet another Prop along with global props. I see your point since the course API call is made anyways after the global call and the term data is present there. If you go through Globals I think it would be making API call to get the course term via course API endpoint and then repeat. 

What you did make sense!

File: ccm_web/server/src/api/api.utils.ts
Comment (ssciolla) [inline ccm_web/server/src/api/api.utils.ts:41]: I think I suggested this initially, but I was seeing some TypeScript errors when I tried to use it. Because the `T` isn't actually used anywhere, I don't think it's essential that we use `extends` to limit what types are accepted here. Basically, it will be `T` if it's not `APIErrorData`. One cool thing is that by removing this, we can actually use this for calls that return multiple objects, since we can pass, for example, `CanvasCourse[]` as the type parameter, and it all still works!

Comment (ssciolla) [reply ccm_web/server/src/api/api.utils.ts:41]: Besides, I am now seeing an error where this function returns `successes` that looks like this:
```
Type 'CanvasEnrollment' is not assignable to type 'T'.
          'CanvasEnrollment' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint 'CanvasEntity'.ts
```
This sent me down a rabbit hole of SO posts and TypeScript docs, and I'm not entirely sure I can explain it. Perhaps TS can't ensure that the same member of the `CanvasEntity` union type is accepted and returned from the function? The error goes away when you remove the `CanvasEntity` constraint.

**Comment by chrisrrowland** [comment]
I started trying to integrate this with my front end PR and noted a couple of items which have been addressed.

The change to CourseWithSections requires some moderate refactoring of front end code so I haven't been able to actually test yet, however eyeballing the code it looks ok to me and I'd perfer just to merge and figure out if it needs more changes later.

File: ccm_web/server/src/api/dtos/api.section.ids.dto.ts
Comment (ssciolla) [inline ccm_web/server/src/api/dtos/api.section.ids.dto.ts:16]: Not sure about the `250` limit here, but it seems responsible to put --some-- limit on this.

File: ccm_web/server/src/canvas/canvas.interfaces.ts
Comment (ssciolla) [inline ccm_web/server/src/canvas/canvas.interfaces.ts:52]: I think we need a `Base` because operations like `merge` and `unmerge` don't have access to `total_students`. Will double check that.

File: ccm_web/server/src/canvas/canvas.interfaces.ts
Comment (ssciolla) [inline ccm_web/server/src/canvas/canvas.interfaces.ts:82]: It was a deliberate choice not to use the `Canvas` prefix here. I thought maybe composite, non-atomic entities might be easier to identify if we didn't use that. Open to changing it though.

**Comment by ssciolla** [comment]
I think this is ready for review. I'm going to look at this Codacy alert again, and also do some more testing. @lsloan, I think @pushyamig and @chrisrrowland can probably handle this, since they've already started looking at it, but I've incorporated some of the refactoring stuff we discussed on PR #165 and there are changes I want you to be aware of. Of course, you can review if you would like to.

**Review by chrisrrowland — APPROVED** [review]
Approving so we can merge and I can more effectively review.

Been trying to get the FE and BE to align to actually test but it's been quite a bit of changes and I'd rather just figure out if something needs to change after merging.

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [inline ccm_web/server/src/api/api.service.ts:123]: What is the thinking behind putting the merge/unmerge sections in separate handlers course/section. But in the course handler you are calling the section handler for unmerging it. I felt for the API with course_id dependency seems to be in course handler and things with sections dependency seems to be section. 

I think for future API call, If you document the intent that it would be easier for follow along the similar pattern

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [inline ccm_web/server/src/api/api.service.ts:73]: @ssciolla I think the sections from the API `/api/v1/accounts/:account_id/courses` from the API will pull all state like `created, claimed, available, completed` . We should document this here just for  clarity in code or swagger API( leave it if adds more work) 

I feel in the testing Melinda can decide which list to populate. Courses in canvas are create 6 month before the term. Who know when an instructor will decide to merge section. I don't think any action is needed here now but sharing some insights

File: ccm_web/server/src/api/dtos/api.section.ids.dto.ts
Comment (pushyamig) [reply ccm_web/server/src/api/dtos/api.section.ids.dto.ts:16]: The use case don't call for bulk create that many section at-least based on todays Canvas usage by Admin staff . 60 is the limit we set and the Backend handles it. But 250 limit seem ok and code can handle it with async calls

File: ccm_web/server/src/canvas/canvas.interfaces.ts
Comment (pushyamig) [reply ccm_web/server/src/canvas/canvas.interfaces.ts:57]: Yeah... it is not returned as part of merge/unmerge sections. 

I think the course/section object in the FE is following the same object pattern as Backend. May be that should reflect this pattern? https://github.com/tl-its-umich-edu/canvas-course-manager-next/blob/main/ccm_web/client/src/models/canvas.ts#L6

File: ccm_web/server/src/api/api.utils.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.utils.ts:41]: I appreciate the effort here. We can leave as is now.

File: ccm_web/server/src/api/api.course.handler.ts
Comment (pushyamig) [inline ccm_web/server/src/api/api.course.handler.ts:113]: @ssciolla Technically you can cross-list across terms. Like Summer 2021 to Spring 2020 etc. 
I tested the use-case from the PR it allows to cross-list and decross list multiple terms from Swagger-UI

I know our workflow only allows to presents sections from same terms. Should be allow more checks if the both course and section belongs to same term?

But this shouldn't be show stopper to merge this. This just needs more thoughts in the BE and user can't so this from the UI

File: ccm_web/server/src/api/api.service.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:123]: This is a good question. Let me try to explain why I did it this way, and then maybe you can give some advice on how to document it for future use.

I put `mergeSection` and `unmergeSection` in the `SectionHandler` because the routes in Canvas are section-oriented, meaning they use `sections/x` as a base, like the other section calls (`sections/${this.sectionId}/crosslist/${targetCourseId}` and `sections/${this.sectionId}/crosslist`).

In the code you commented on, I chose to only use `SectionApiHandler`s for `unmergeSections` because there isn't a need to identify a course in that operations (see above route). For `mergeSections`, I could have done something similar and not used `CourseApiHandler` at all, but I thought it might be convenient to have a method on that class that would merge multiple sections to one course. I used `SectionApiHandler` in that method since the operation for merging was already defined.

Hopefully that helps explain my logic? If you'd prefer, I can move the `CourseApiHandler.mergeSections` code to `APIService.mergeSections`, I'm not super attached to one implementation over another.

**Comment by pushyamig** [comment]
>seems like everyone can access all routes, they just may not get data back/Consider limiting access to /admin/sections/ by role from LTI launch (probably a separate issue)

You are bringing some interesting insights that we never focused before. Yeah right now FE is the only place we check for the role and prevent feature access. With scoped token they won't see any data that they don't access to it. 

We can have a discuss if this is important for first phase, since we are guarded from incorrect access of data, But it's more making BE check explicitly  again.

File: ccm_web/server/src/api/api.service.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:73]: Okay, would you like me to add a comment about the states returned by default? I can see maybe removing `completed` from that list, otherwise, I think the others are probably okay. Maybe we should ask Melinda?

File: ccm_web/server/src/api/api.course.handler.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.course.handler.ts:113]: We could limit that, although it would probably involve more API calls to check the course and section.

I guess my thinking at this point is that if the user has modify permissions for the course and section (which scoped permissions will enforce), there's not technically any harm to keeping this possible. They would be able to do this through the API if they minted their own token and hit the Canvas API directly. Besides, I think it would be non-trivial for a user to hit our API directly, they'd have to find the API routes and the CSRF token.

File: ccm_web/server/src/canvas/canvas.interfaces.ts
Comment (ssciolla) [reply ccm_web/server/src/canvas/canvas.interfaces.ts:57]: Yes, good point, I can change that. Hopefully it doesn't mess with @chrisrrowland too much.

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts:73]: @melindakraft What are your thoughts on this?

File: ccm_web/server/src/api/dtos/api.section.ids.dto.ts
Comment (ssciolla) [reply ccm_web/server/src/api/dtos/api.section.ids.dto.ts:16]: This would be for merging and un-merging sections. I think I will leave it as is for now, but we can modify it later as needed.

There may be some decorators I'm using here that we should add to the `CreationSectionsDto`, like the `ArrayUnique` and `ArrayMaxSize`. Separate issue though.

File: ccm_web/server/src/api/api.course.handler.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.course.handler.ts:113]: OK, We can leave it for now and document it

File: ccm_web/server/src/api/api.service.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:73]: Maybe `completed` isn't really a problem since we're limiting by term. While I don't think people would really want to cross-list for past terms (where you'd have `completed` courses), I could see someone doing that for testing. In general, you wouldn't see too many `completed` courses in the current or future terms, so it wouldn't be confusing overly confusing. Also, I could imagine an instructor "complete"ing a course accidentally (instead of keeping it un-published) and still want to crosslist sections from it.

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts:123]: > I can move the CourseApiHandler.mergeSections code to APIService.mergeSections

Don't need to move anything. Document it saying that based on the canvas API routing respective API call are placed in respective handler, that's should be good enough for understanding.

File: ccm_web/server/src/api/dtos/api.section.ids.dto.ts
Comment (pushyamig) [reply ccm_web/server/src/api/dtos/api.section.ids.dto.ts:16]: >This would be for merging and un-merging sections. I think I will leave it as is for now, but we can modify it later as needed.

Sorry I missed noticing that. In that case LIMIT is OK.

May be this dto can have a name like `api.merge.unmerge.section.id's.dto.ts`?

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts:73]: Yes. I agree on the accidentally marking Conclude the course. I don't know conclude or complete same thing. But We can show the completed course. No harm. IMO

File: ccm_web/server/src/api/dtos/api.section.ids.dto.ts
Comment (ssciolla) [reply ccm_web/server/src/api/dtos/api.section.ids.dto.ts:16]: I prefer to leave it as general purpose/general naming, since there's nothing about it that is limited to those use cases. If you want me to add a comment indicating where it is used, I can do that.

File: ccm_web/server/src/canvas/canvas.interfaces.ts
Comment (ssciolla) [reply ccm_web/server/src/canvas/canvas.interfaces.ts:57]: I added a bit more typing to the client `models/canvas.ts` (see https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/199/commits/a014d63d6431cd0621d27e802816c06fb7a44a3e) to reflect the server changes.

I'm going to leave the `api.ts` definitions to Chris, since he's already done some of that work and may have opinions on what parameters they accept, how they are named, but @chrisrrowland, please note the nuance that the merge and unmerge routes do not return `total_students`, so you may need to hold the original section objects in state and update the `course_id` (or do some other client magic).

File: ccm_web/server/src/api/api.course.handler.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.course.handler.ts:113]: I think we ever open the Swagger API to Prod environment for Functional team, then I think we need to think back on all these things.

**Review by pushyamig — APPROVED** [review]
This Looks Good. There some documentation requested, other that the PR flow for Merge section feature is good. I tested all the case around it.

The refactoring also looks ok, done to fit the new API calls for Merge section.  I did not tested the API endpoint with refactored code, but should work as these are just code structuring and moved the methods based on that.

The termId is sent to FE via course API and FE should wire that by passing to respective components @chrisrrowland

File: ccm_web/server/src/api/api.service.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:73]: Okay, @pushyamig, see comment for requested documentation: https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/199/commits/774571d19053c4057a3bfb1ba9ede1e4c7bcc97f

File: ccm_web/server/src/api/api.course.handler.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.course.handler.ts:117]: I'm not sure we'll ever want to do that, but we can discuss.

Comment (pushyamig) [reply ccm_web/server/src/api/api.course.handler.ts:117]: I am not saying opening the swagger UI for this first phase. I don't want to give the access to even dev other than local development. 

But there was some discussion around this early on among dev making it available in Prod for future. So I have to mention this now from that perspective.

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts:73]: Thanks for the comments

Comment (melindakraft) [reply ccm_web/server/src/api/api.service.ts:73]: @ssciolla @pushyamig For what it is worth, when a course is concluded, all enrollments are archived, including the instructor's access (technically, they are still there but in an inactive state). According to Canvas, I think we can determine that completed = concluded: https://community.canvaslms.com/t5/Instructor-Guide/What-are-the-different-states-of-a-Canvas-course/ta-p/1184

I don't think we should show concluded/completed courses for the purpose of section merging. If a course is accidentally concluded, there will be no section or enrollments. it is easy for an admin in Canvas to unconclude a course, though that won't put the enrollments back. That step has to be done separately. I would not expect CCM to fill that gap since it is an edge case. 

Regarding crosslisting across terms, it comes up only occasionally and can be handled outside of CCM. It is an edge case, too.

File: ccm_web/server/src/api/api.course.handler.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.course.handler.ts:117]: Gotcha. Was there something else I needed to do here documentation wise?

Comment (pushyamig) [reply ccm_web/server/src/api/api.course.handler.ts:117]: I wanted to document, that BE allows for cross-listing through swagger UI. But the FE flow disallows for user from UI

File: ccm_web/server/src/api/api.service.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:123]: Cool! There is more we could do with the handler refactoring. We could probably add an intermediary handler class for accounts (`AccountApiHandler`) and a non-admin handler (instructor/student), but I didn't see a huge benefit to those (at least not yet). We could also add abstract base classes, do other OOPsy stuff, but we can explore all this later if we wanted to build a more robust API. Also, GraphQL could change this significantly, whenever that rolls around. I will open a separate issue to capture the ideas here, but it's definitely Icebox material.

File: ccm_web/server/src/api/api.course.handler.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.course.handler.ts:117]: K, not sure where this should be documented, in code or not. I'll leave it as for now.

File: ccm_web/server/src/api/api.service.ts
Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:73]: Okay, let me add the state parameter then for everything but `deleted` and `completed`, and test before merging. Thanks for the input!

Comment (ssciolla) [reply ccm_web/server/src/api/api.service.ts:73]: Okay, see https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/199/commits/3c03dc6314c08fd7ef15d3b2bb7b6070c809a63f

**Comment by ssciolla** [comment]
Okay, I'm merging now, thanks for the reviews! If you see any bugs, please point them out to me, or create an issue and assign me.

File: ccm_web/server/src/api/api.service.ts
Comment (pushyamig) [reply ccm_web/server/src/api/api.service.ts:73]: The new code make sense, did not test though

**Pull request merged**
